Refactor selections with non-overlapping data structure

detarkende opened the pull request:
<!-- DO NOT POST LINKS OR REFERENCES TO COPYRIGHTED CONTENT IN YOUR ISSUE. -->

**What is the purpose of this pull request? (put an "X" next to item)**

- [X] Documentation update
- [X] Bug fix
- [ ] New feature
- [ ] Other, please explain:

**What changes did you make? (Give an overview)**

### WIP - please give me feedback

The goal of this PR is to introduce a new data structure that handles piece selections. The new `Selection` class allows you to add intervals to the selection and remove them without creating overlapping selection items.

Upon insertion, the class checks if the code overlaps with any existing selections. If it does, then it merges the two selections (the new, inserted selection remains intact and the existing one(s) are truncated or cut into pieces).

This also means that (for example) if the user runs
```js
torrent.deselect(0, torrent.pieces.length - 1);
```
then all the pieces are actually deselected.

The long-term goal of this change is to also fix the issue with `file.deselect()` but that's a story for a later day.

**Which issue (if any) does this pull request address?**

It relates to #2755.

**Is there anything you'd like reviewers to focus on?**

As Webtorrent is a pretty large project, I don't understand all of it. I would like reviewers to help me poke holes into this implementation. If there are any aspects of the app that this breaks that I didn't think of, please let me know so I can hopefully extend my solution.

detarkende committed:
Refactor selections with non-overlapping data structure

ThaUnknown commented:
code quality is good, I haven't look at the actual functionality tho, it would be good to actually explain what you did, and how the algorithm changed, you also cannot get rid of priority.

ThaUnknown commented:
also if you're already addessing select/deselect behavior it would be good if you included what <https://github.com/webtorrent/webtorrent/pull/2115> tries to accomplish, not exactly its code, but it's functionality

detarkende commented:
## Explanation of my changes:

To make the _selections array non-overlapping, I abstraced it away into a separate class.
This class internally has the same array, but its `insert` and `remove` methods include some extra logic to keep items separate:

When we remove an interval, we loop over the existing items and check if there's any overlap with the removed interval. There can be 4 types of overlapping:
- `isLowerIntersecting`: Only the lower end of the removed interval overlaps 
    - Example: we are removing `8-12`, but `1-10` is already selected, so we cut the end of the existing selection off.

- `isUpperIntersecting`: Only the upper end of the removed interval overlaps 
    - Example: we are removing `10-22`, but `20-25` is already selected, so we cut the beginning of the existing selection off.
 
- `isInsideExisting`: The removed interval is specifically inside an existing larger selection. 
    - Example: we are removing `10-15`, but `1-30` is already selected, so we carve out the removed section from the middle of the existing one -> this leaves us with 2 selections with a hole between them.

- `isCoveringExisting`: This is the simplest case. If the removed interval covers an entire existing selection, or if they're an exact match size-wise, then we simply remove the existing selection.

Insertion is very simple compared to this: we call the `remove` method with the newly inserted interval to make sure that there's no overlapping items, then we push the selection to the array.

---

### On the topic of priorities:

I understand why we assign priorities to pieces when they are selected, but I don't see why we would need them upon deselection. 

From recently reading around I got the feeling that it's for making sure that we don't deselect intervals that were selected by file streams -- therefore in the past we would only deselect the given pieces if the start and end points matched exactly and also the priority. 

Is this right? If so, then wouldn't it make more sense to distinguish stream-selections from non-stream selections with a boolean flag instead of their priority? If this is the case, then I'll adjust my code to make sure that only files can deselect file streams and not end users accidentally. Let me know if I'm wrong on this.

ThaUnknown commented:
> I understand why we assign priorities to pieces when they are selected, but I don't see why we would need them upon deselection.

your assumption is roughly correct, but it doesn't apply to only streams, imagine webtorrent's HTTP server being used as a media server, and having say 10 people watching 1 video as it's being downloaded/streamed from BitTorrent

if multiple people are at the same point in time in the video, and one of them stops watching it unloads the video, which sends a cancel request to the range that was requested, thus deselecting the pieces for ALL 9 other people that are watching it, permanently stalling the playback for 9 people until someone reloads it

this ofc applies to not only streams, but manually running selection ranges, your code works for single selections, but when you have multiple unrelated processes making selections, it falls apart

hope that clears it up

detarkende commented:
I think I finally understand. So the file stores one iterator for each stream created. If there are multiple requests as the same time, there might be 2 iterators for the same (or overlapping) piece intervals. When the file iterator's `destroy` method is called, it deselects the selected pieces. Which means that if selections cannot overlap, then one iterator might clear its own selection, but with that also clear the selection for all other iterators.

I think I might be able to solve this by keeping a separate array of selections specifically for these cases. This would ensure that existing features don't break. I'll try to write the code now and push it as soon as I can. I'll try to setup an example project as well, to test that my solution actually works.

ThaUnknown commented:
yeup, u got it right.

detarkende committed:
fix: Add stream flag to selections that follow old select/deselect behavior

detarkende committed:
feat: add option to create torrent with no pieces selected

detarkende committed:
fix: fix torrent selection bug and handle deselected start

detarkende committed:
fix: update api.md

detarkende committed:
fix: refactor selection test toString function

detarkende commented:
@ThaUnknown Thanks for the guidance so far. I have implemented 2 things:
1. Streams create specific stream-selections, which are marked with a true `isStreamSelection` flag. These can only be deselected by streams, and only one at a time. The actual implementation of this is mostly the same as in #2115, since that has already received a round of code review - so I assume it's not a terrible way to do it 😄 
2. In #2115, @feross suggested a new option in the torrent options: the `deselect` option. I added that as well.

I copied the tests from Alex's PR and modified them somewhat + just to be on the safe side, I created a test server and connected to it with 2 clients, then disconnected one of them. The other client's selection remained in the `_selections` and its video playback continued without any issues.

I read somewhere that you should never modify the git history, so I basically sent everything in a separate commit, let me know if I should squash my commits though.

ThaUnknown commented on lib/torrent.js:
```diff
@@ -35,6 +35,7 @@ import utp from './utp.cjs' // browser exclude
 import WebConn from './webconn.js'
 
 import info from '../package.json' assert { type: 'json' }
+import { Selection } from './selection.js'
```
this should be above the info import, right under webconn

detarkende replied:
got it, thanks. fixed it

ThaUnknown commented on lib/torrent.js:
```diff
@@ -688,7 +690,9 @@ export default class Torrent extends EventEmitter {
     this.pieces[index] = new Piece(len)
     this.bitfield.set(index, false)
-    this.select(index, index, 1)
+    if (!this._startAsDeselected) {
```
inline this without {}

detarkende replied:
done.

ThaUnknown commented on lib/torrent.js:
```diff
@@ -1019,7 +1025,7 @@ export default class Torrent extends EventEmitter {
   }
 
-  select (start, end, priority, notify) {
+  _select (start, end, priority, notify, isStreamSelection) {
```
notify = noop, isStreamSelection = false) {

detarkende replied:
fixed.

ThaUnknown commented on lib/torrent.js:
```diff
@@ -1029,36 +1035,37 @@ export default class Torrent extends EventEmitter {
+    }, isStreamSelection)
 
-    this._selections.sort((a, b) => b.priority - a.priority)
+    this._selections.sort((a, b) => a.priority - b.priority)
```
why was the order reversed?

detarkende replied:
honestly no clue. probably just an accident, but good catch. I reverted to the old order

ThaUnknown commented on lib/selection.js:
```diff
@@ -0,0 +1,202 @@
+ * @property {boolean} isStreamSelection
+ */
+
+export class Selection {
```
this should be Selections plural, since it handles many selections

ThaUnknown commented on lib/selection.js:
```diff
@@ -0,0 +1,202 @@
+export class Selection {
+  /**
+   * @type {Array<SelectionItem>}
+   * @private
```
skip these @private and @public jsdocs, we'd use # if we wanted to declare private, but we don't because we want people to access internal functions if they wish to, which is why _ is used for private, for strict engines such as TS, touching this will throw errors, which is annoying

detarkende replied:
You're right, thanks for this. I removed them all

ThaUnknown commented on lib/selection.js:
```diff
@@ -0,0 +1,202 @@
+  /**
+   * @public
+   */
+  isEmpty () {
```
this is VERY redundant

detarkende replied:
can't really argue with that haha 😄 removed it.

ThaUnknown commented on lib/selection.js:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  * [Symbol.iterator] () {
+    const self = this
```
no.... don't use self
do remove: () => {}

detarkende replied:
thanks, good idea. fixed.

ThaUnknown commented on lib/selection.js:
```diff
@@ -0,0 +1,202 @@
+    const self = this
+    for (let i = 0; i < this._items.length; i++) {
+      const item = this._items[i]
+      yield {
```
this.... works? but feels a bit too hacky? if u figure out a cleaner way to do this then great, if not oh well it's fine

detarkende replied:
I don't consider this to be too hacky to be honest. The other solution would be to call `.splice` on `_selections._items` from outside of this file, which does feel hacky tbh.
Plus this allows us to use `for (.. of ..)` loops and still remove items without having to deal with the indexes. If you think of a better solution, I'd be happy to replace the current one, but I also don't really see this as an issue either, since it's easily understandable and simple.

ThaUnknown commented on lib/selection.js:
```diff
@@ -0,0 +1,202 @@
+ */
+export function isCoveringExisting (newItem, existing) {
+  return newItem.from <= existing.from && newItem.to >= existing.to
+}
```
you don't need return types on any of these as tslib will detect that this always returns bool, the returns comments you could move into the bottom of the top comment

ThaUnknown replied:
these params should also be `{from: number, to: number}` or `MinimalSelectionItem` if you want to make something like that, not SelectionItem

detarkende replied:
The return types are not there necessarily there to provide the types, but rather to leave a comment on **when** this type returns true. As you can see they all have a _"returns true if the new items..."_ explanation after the `@returns {boolean}`. To me it makes sense to put these comments to the `@returns` section, but let me know if you'd still like me to change it.

detarkende replied:
Also, I like the idea of having a smaller `MinimalSelectionItem` typedef, so thanks. I added that too.

ThaUnknown commented on lib/torrent.js:
```diff
@@ -1029,36 +1035,37 @@ export default class Torrent extends EventEmitter {
     this._debug('select %s-%s (priority %s)', start, end, priority)
 
-    this._selections.push({
+    this._selections.insert({
```
SelectionItem implements `isStreamSelection` so simply pass it in the object, rather than creating a new function parameter?

detarkende replied:
That's a good point. I removed the extra argument from `insert` and `select`, `deselect`. It's now determined by the shape of the object.

ThaUnknown reviewed (changes requested):
This PR is high quality, so I'm going to be VERY strict on it :)

hope that doesn't demotivate you

detarkende commented:
I hope so too. Sure, go for it 😄

ThaUnknown commented:
Webtorrent's selections also have 1 more issue, I'm not gonna ask you to fix it, but it would be insanely welcome as it's the last thing that sucks in this implementation: webtorrent selects full pieces, rather than specific bytes.

This becomes an issue for example when there's a torrent with 2 files, and file 1 ends at 1/2 piece and file 2 starts at where that 1/2 piece ended, selecting and deselecting the files, impacts the other file's selections.
this is especially an issue when creating file streams/iterators, as in theory removing an iterator from the start of the 2nd file, can remove the selection for the end of the 1st file, which ignores metadata download, and effectively stalls all playback etc

The way I could see this solved, is that internally we'd select byte ranges, but users would still request pieces.
This means that file selections aka torrent.file.select would select ranges, and deselecting it would also deselect ranges, then the indexes specific selections target would be exposed as parts of those ranges.

This is hard to explain but TLDR: internally selections should be byte ranges, however they should be exposed to users and the selection algorithms as piece numbers.

If you're interested in this, and didn't understand my shit explanations, ask, and I'll put more effort into properly explaining this.

ThaUnknown commented on lib/selection.js:
```diff
@@ -0,0 +1,202 @@
+   */
+  remove (item, isStreamSelection = false) {
+    for (let i = 0; i < this._items.length; i++) {
+      const existing = this._items[i]
```
you do not check priorities here, remove should remove by priorities!!!!!!!!!!!

detarkende replied:
I added the priority check to the section that compares `isStreamSelection` selections

For regular `select` and `deselect`, priority shouldn't matter:
- If it's an insertion, then it should overwrite whatever was there previously.
- If we are using `deselect`, then the user shouldn't have to know beforehand what the current selection's priority is. Especially when you deselect a range that contains multiple existing ranges - they could all be different selections.

detarkende committed:
fix: libs/torrent import order

detarkende committed:
fix: inline if statement in markUnverified

detarkende committed:
fix: torrent.select default arguments

detarkende committed:
fix: reversed sorting at end of torrent.select

detarkende committed:
fix: rename selection class to plural - selections + filenames

detarkende committed:
fix: remove @private and @public tags from jsdoc comments

detarkende committed:
fix: remove  redundant selections.isEmpty method

detarkende committed:
fix: replace `self` usage with arrow function to preserve `this`

detarkende committed:
fix: create new MinimalSelectionItem typedef and remove examples from jsdoc

detarkende committed:
fix: use isStreamSelection from object instead separate param and fix SelectionItem typedef

detarkende committed:
fix: add priority check to selections.remove when isStreamSelection === true

detarkende commented:
> Webtorrent's selections also have 1 more issue, I'm not gonna ask you to fix it, but it would be insanely welcome as it's the last thing that sucks in this implementation: webtorrent selects full pieces, rather than specific bytes.

Personally, I like the idea of having byte-range selections. I haven't given this too much thought yet, but I imagine it could be used to specifically not download any files, other than the selected one(s). For me this was an issue previously because when a torrent has 2 large files and there's a piece that contains data from both files, it creates the second file too (even when I haven't selected it). Now this shouldn't be an issue, since I didn't actually download the second file's contents, but to the file system it is indicated that the second file is also a full file (let's say 10GB). Now I know it's not actually taking up that space, but it's still annoying.

That said, I think that would be a pretty big undertaking and considering that this PR has also grown to be around ~450 LOC in changes, I would prefer to keep those changes in a separate PR.

I can't guarantee to work on it after this PR, because I would first like to address some other issues - namely fixing the very out of date `@types/webtorrent` declaration files, or maybe adding more jsdoc annotations to the code, but I might give it a go later.

TLDR: I like the idea and support it, but I think it should be in a separate PR. I won't start working on it right away, but if I have time later, I might take a stab at it.

detarkende commented:
I'm going to check out the failing tests, but I adressed your comments. 

Thank you for the detailed review! I don't think there are too many details left to debate, so we might ship this thing quite quickly 🚀 _Who said open-source is slow?_ 😄

ThaUnknown commented on lib/selections.js:
```diff
@@ -152,35 +134,26 @@ export class Selection {
- * new_item: 8-12 | 10-15
- * @param {SelectionItem} newItem
- * @param {SelectionItem} existing
+ * @param {MinimalSelectionItem} newItem
```
aw man those explanations were good :(

detarkende replied:
you mean the examples? I can add them back, I just thought it made things kind of messy, but if you found them helpful, then they're probably worth keeping

ThaUnknown replied:
for you maybe, but for some1 who hasnt worked on this in the future it's useful

detarkende replied:
okay, sure. I added them back in then :)

ThaUnknown commented:
> but I think it should be in a separate PR

very fair

> Who said open-source is slow? 

it is, I just sacrifice a lot of free time for FOSS, as you see the previous PR is still open

detarkende committed:
fix: readd examples for interval comparison functions

ThaUnknown commented on lib/torrent.js:
```diff
@@ -1029,36 +1033,38 @@ export default class Torrent extends EventEmitter {
-        break
-      }
-    }
+    this._selections.remove({ from: start, to: end, isStreamSelection })
```
:/ this throws typeerror during static analysis
```
Argument of type '{ from: any; to: any; isStreamSelection: boolean; }' is not assignable to parameter of type 'SelectionItem'.
  Type '{ from: any; to: any; isStreamSelection: boolean; }' is missing the following properties from type '{ offset: number; priority: number; notify: Function; isStreamSelection?: boolean | undefined; }': offset, priority, notifyts(2345)
```

detarkende replied:
Thanks, I updated it to 
```js
/**
* @param {MinimalSelectionItem & {isStreamSelection: boolean, priority?: number}} item 
*/

ThaUnknown reviewed (changes requested):
yup, this is gonna be a breaking change, but I think it's needed and I don't know if its the best idea, @SilentBot1 what do u think

ThaUnknown commented on test/client-deselect.js:
```diff
@@ -0,0 +1,109 @@
+    client2.add(parsedTorrent, { store: MemoryChunkStore, ...addTorrentProps }, (torrent) => {
+      onTorrent(torrent)
+
+      torrent.addPeer(`localhost:${client1.torrentPort}`)
```
you're adding a peer by IP, this is a no-go in browser, move this test to test/node

detarkende committed:
fix: move client-deselect test to node tests

detarkende committed:
fix: update remove method parameter jsdoc

detarkende commented on lib/file.js:
```diff
@@ -91,7 +91,7 @@ export default class File extends EventEmitter {
 
   deselect () {
```
I replied below: https://github.com/webtorrent/webtorrent/pull/2757#discussion_r1511719326

ThaUnknown commented on lib/torrent.js:
```diff
@@ -1029,36 +1033,38 @@ export default class Torrent extends EventEmitter {
     this._updateSelections()
   }
 
+  deselect (start, end) {
```
sorry, there's some overlap between what I want and find missing in WebTorrent, and what you're doing, so I partially wanted you to do it. Selections should be unique by priority levels, imagine a torrent client UI, where u can download different files, at different priorities, but if you want to remove that elevated priority from a file, you'd deselect the file.... instead of removing the selection at that priority

will adding a new selection at a lower priority for only part of an older higher selection override only the overlapping part?

I hadn't had much time to delve into the code, so I was making rough assumptions by quick glances

if that's the case then the if  `if (existing.from === item.from && existing.to === item.to && existing.priority === item.priority)` where I asked u to add priorities is stupid and it was simply my lack of understanding

ThaUnknown replied:
![image](https://github.com/webtorrent/webtorrent/assets/6506529/a96e2104-65a6-47fb-9bcb-d73238872d78)
@detarkende my wonderful paint skills

ThaUnknown commented on lib/selections.js:
```diff
@@ -0,0 +1,184 @@
+      if (existing.isStreamSelection) {
+        if (item.isStreamSelection) {
+          // If both are stream selections and they match, then we remove the first matching item, then we break the loop
+          if (existing.from === item.from && existing.to === item.to && existing.priority === item.priority) {
```
i was stupid, this priority check isnt needed

detarkende replied:
Are you sure? This check is for stream selections only. Those are still on the old logic: from, to and prio have to match and if you have multiple selections for the same range, then you have to call deselect for each.

ThaUnknown replied:
but stream selections cannot be created with priorities, unless some1 does it manually ,which isn't it's intended use case? or am I overlooking smth

ThaUnknown replied:
sorry i mean the if statement stays, but the priority check is useless

detarkende replied:
Oh, okay. I didn't think it through and thought there might be stream selections with differing priorities, but that's not the case. In that case, I removed the priority comparison from the if statement.

ThaUnknown replied:
ah no, streams simply select needed data with the default unset priority, and data that's needed for the stream load RIGHT NOW is set to critical, which is highest priority

ThaUnknown commented on lib/selections.js:
```diff
@@ -0,0 +1,184 @@
+   * @param {MinimalSelectionItem & {isStreamSelection: boolean, priority?: number}} item Interval to be removed from the selection
+   */
+  remove (item) {
+    for (let i = 0; i < this._items.length; i++) {
```
should this be a `for (const existing of this) {` then later use either existing.remove() or this.swap()?

ThaUnknown replied:
nvm it wouldn't be swap, but remove could be used

detarkende replied:
we also need `i` for the `splice` call on line 55.

detarkende committed:
fix: remove priority checks for stream selection remove

ThaUnknown commented on test/node/client-deselect.js:
```diff
@@ -0,0 +1,109 @@
+test('client.select: partial torrent', function (t) {
+  t.plan(3)
+
+  let lastPiece
```
this is touch missleading, torrent's have a concept of "lastPiece" which is "the last piece in the torrent which doesn't have the length of a normal piece, instead is the remaining data" so this could be treated as lastPieceLength, better be safe and rename this to `lastPieceIndex` everywhere in this test file

detarkende replied:
didn't know about that, plus this test is not originally mine. thanks for the info. renamed them all

detarkende committed:
fix: rename lastPiece to lastPieceIndex in tests

ThaUnknown commented on lib/torrent.js:
```diff
@@ -549,7 +551,7 @@ export default class Torrent extends EventEmitter {
       // start off selecting the entire torrent with low priority
-      if (this.pieces.length !== 0) {
+      if (this.pieces.length !== 0 && !this._startAsDeselected) {
         this.select(0, this.pieces.length - 1, false)
```
what about this, why false?

detarkende replied:
I didn't change this line, but I guess in this case I could replace the `false` to `0`. They're basically the same thing here.
Replaced in last commit

ThaUnknown reviewed (commented):
One issue stands with these changes in total, `notify` can be swallowed or overwritten, that's kinda bad.... when you merge or overwrite selections, you should also merge their notify functions? imagine multiple functions ask for notify on different priority levels, all but one will be lost....

detarkende committed:
fix: replace select priority boolean with number

ThaUnknown commented:
yeah @detarkende you'll need to "merge" callback/notify functions when selections are merged/overwritten, this will call the notify function when it's not really necessary, but its better to call it more than less, I think it would be roughly like this:
```js
const existingStart = { ...existing, to: item.from - 1, notify: ()=>{item.notify();existing.notify()} }
```
probably? you'd need to double check this, and ofc it would be in other places too

SilentBot1 commented:
The code looks overall high quality, thanks for taking the time to PR this @detarkende.

Through my testing, I only found one real issue, and one thing that may cause issues:

1. The first being that the `&so=` magnet query parameter is preventing any selections from being made:
```js
client.add('magnet:?xt=urn:btih:08ada5a7a6183aae1e09d831df6748d566095a10&dn=Sintel&tr=udp%3A%2F%2Fexplodie.org%3A6969&xs=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel.torrent&so=0', {deselect: false})
// No selection made - incorrect

client.add('magnet:?xt=urn:btih:08ada5a7a6183aae1e09d831df6748d566095a10&dn=Sintel&tr=udp%3A%2F%2Fexplodie.org%3A6969&xs=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel.torrent&so=1')
// No selection made - incorrect

client.add('magnet:?xt=urn:btih:08ada5a7a6183aae1e09d831df6748d566095a10&dn=Sintel&tr=udp%3A%2F%2Fexplodie.org%3A6969&xs=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel.torrent&so=5', {deselect: true})
// No selection made - I would say this is correct, explicit deselect should take priority over &so=
```

2. The second being that two overlapping sections at the start of the piece range leaves a negative selection:
```js
client.torrents[0].select(0, 100)
client.torrents[0].select(0, 50)
```
which leads to an unfulfillable selection:
```js
Selections {
  _items: [
    {
      from: 0,
      to: -1,
      offset: 0,
      priority: 0,
      notify: [Function: noop],
      isStreamSelection: false
    },
    ...
  ]
}
```

Other than that, it looks like a nice addition!

detarkende commented:
> yeah @detarkende you'll need to "merge" callback/notify functions when selections are merged/overwritten, this will call the notify function when it's not really necessary, but its better to call it more than less, I think it would be roughly like this:
> 
> ```js
> const existingStart = { ...existing, to: item.from - 1, notify: ()=>{item.notify();existing.notify()} }
> ```
> 
> probably? you'd need to double check this, and ofc it would be in other places too

@ThaUnknown Thanks for the idea. I'm currently trying to make it work with an independent notifications array inside Selections. It's kinda the same idea what #2115 does. I have some issues with it atm, but I'll try to fix them. If I can't I'll probably use your idea.

detarkende committed:
fix: handle notify functions the old way with separate notification store

detarkende commented:
@ThaUnknown pushed some changes that address the notifications issue. I believe it solves the problem, but maybe I didn't think of some edge case. lmk what you think.

detarkende committed:
fix: issues with so and deselect flags

detarkende committed:
fix: when reducing from by 1, handle 0 edge case (shouldn't go below 0)

detarkende commented:
@SilentBot1 Thank you for the review. You've got a good eye for bugs, I would've never caught those two by myself.
I addressed both of them in my last 2 commits. Let me know what you think

ThaUnknown commented on lib/selections.js:
```diff
@@ -0,0 +1,222 @@
+   * @param {NotificationItem} notification
+   */
+  removeNotification (notification) {
+    const index = this._notifications.findIndex(n => n === notification)
```
.indexOf()

ThaUnknown commented on lib/selections.js:
```diff
@@ -0,0 +1,222 @@
+      this.remove({ from, to, isStreamSelection })
+    }
+    this._items.push({ from, to, isStreamSelection, ...restItem })
+    this._notifications.push({ from, to, notify })
```
I don't like this, you're essentially using the old range behavior, and it has it's own separate state, not like the rest, it's way too complex

ThaUnknown commented on lib/torrent.js:
```diff
@@ -1302,35 +1306,38 @@ export default class Torrent extends EventEmitter {
 
-      if (oldOffset !== s.offset) s.notify()
+      if (oldOffset !== s.offset) {
+        this._selections.getNotificationsByRange(s.from, s.from + s.offset).forEach(({ notify }) => {
```
this... is gonna be insanely slow when there are a lot of selections?, also use for of, not forEach

ThaUnknown commented:
I'm not a fan of this notification code, because it's complexity skyrocketed compared to the other things, and I can't review it by simply looking at it, and I don't have the time to test it

ThaUnknown commented on lib/selections.js:
```diff
@@ -0,0 +1,222 @@
+  getNotificationsByRange (from, to) {
+    const range = { from, to }
+    return this._notifications.filter(notification => {
+      const isNotEmpty = (notification.to - notification.from) > 0
```
```suggestion
      // isNotEmpty
      if ((notification.to - notification.from) > 0) return false
```

ThaUnknown committed:
chore: simply code logic
fix: notify not emitting for first pieces

ThaUnknown committed:
perf: don't create empty notifications

ThaUnknown commented:
@detarkende as expected, the notification code is too complex and already broken somehow, in a way I can't find, I fixed it not emitting notify for first pieces, made it not run when no value is specified by removing noops, but silent found another issue where removing a manually created selection didn't remove it's notification? there's fuckery afoot

ThaUnknown commented:
also, you made me put actual effort into this, I hope you're proud of yourself..... should tell you that this is an actually important change that's been a long standing issue :p

ThaUnknown commented:
couldn't we get rid of offset like in this PR? https://github.com/webtorrent/webtorrent/pull/2115/files#diff-43f8f193b859b154ddc4cb89276649d35e2eb59859b68eae0c2ad58f7306aa59R1211

ThaUnknown committed:
fix: improve removenotification code

ThaUnknown commented:
silent will test this PR, look into that offset thing, and then we'll merge this if there are no issues

detarkende commented:
> @detarkende as expected, the notification code is too complex and already broken somehow

You seem unsatisfied with the current solution. Do you think we should handle notifications alongside selection ranges and merge them, when we merge the selections? 

I'm definitely not against that, but I also saw that you committed changes to the current version, so now I'm hesitant to change it to your initial idea.

Do you want to fix the current version, or scrap it and move to a different solution?

detarkende commented:
> couldn't we get rid of offset like in this PR? https://github.com/webtorrent/webtorrent/pull/2115/files#diff-43f8f193b859b154ddc4cb89276649d35e2eb59859b68eae0c2ad58f7306aa59R1211

I tried getting rid of the offset, but ran into a bunch of small bugs that were difficult to debug. Not sure what causes them, I was working on something else at the time.

I can look into it, but I personally kinda like the current solution, where we keep the original selection's size and the progress separate.

ThaUnknown commented on lib/torrent.js:
```diff
@@ -1302,23 +1306,28 @@ export default class Torrent extends EventEmitter {
 
-      if (oldOffset !== s.offset) s.notify()
+      if (oldOffset !== s.offset) {
+        for (const { notify } of this._selections.getNotificationsByRange({ from: s.from, to: s.from + s.offset })) {
```
question: shouldnt this be notificationsIntersecting?

ThaUnknown commented on lib/torrent.js:
```diff
@@ -1302,23 +1306,28 @@ export default class Torrent extends EventEmitter {
-      s.notify()
+      s.remove() // remove fully downloaded selection
+      // remove notifications when we remove selection.
+      for (const notification of this._selections.getNotificationsByRange(s)) {
```
and shouldnt this be notificationsInside?

the issue I'm seeing is that notifications which are touching, but arent containing a range, will get removed too, meaning they won't emit for other bigger ranges?

ThaUnknown commented:
> You seem unsatisfied with the current solution. Do you think we should handle notifications alongside selection ranges and merge them, when we merge the selections?

yeah quite so, as you can see there are already issues and uncertainties....

ThaUnknown commented on lib/selections.js:
```diff
@@ -0,0 +1,221 @@
+    if (!isStreamSelection) {
+      this.remove({ from, to, isStreamSelection })
+    }
+    this._items.push({ from, to, isStreamSelection, ...restItem })
```
same here, if its a stream selection, push it to _streamItems not _items

detarkende commented:
I don't have enough time today or tomorrow, but I'll come back to it on Friday and hopefully fix everything that's left. Thank you for your patience.

ThaUnknown commented:
silent found an issue in this, but hasnt reported it, idk why
```js
> client.torrents[0].select(0, 1514, 1, ()=>{})
undefined
> client.torrents[0]._selections
Selections {
  _items: [
    {
      from: 0,
      to: 1514,
      isStreamSelection: false,
      offset: 0,
      priority: 1
    }
  ],
  _notifications: [ { from: 0, to: 1514, notify: [Function (anonymous)] } ]
}
> client.torrents[0].deselect(0, 1514)
undefined
> client.torrents[0]._selections
Selections {
  _items: [],
  _notifications: [ { from: 0, to: 1514, notify: [Function (anonymous)] } ]
}
```

detarkende committed:
fix: storing notifications the overlapping way

ThaUnknown commented:
yeaaaaaaah, this is what I meant with insane complexity

from my POV: 80% of the PR and 90% of the complexity is now the selection notifications, which are rarely gonna be used in the first place
it's complexity is now high enough where I don't think I'll be able to understand and review all the potential cases

I really think we should have some simpler system like merging, and introduce higher accuracy for noti in a separate PR

detarkende commented:
@ThaUnknown, look, I agree with you. I don't like that notifications is this complex. I wouldn't say that it's 80% of this PRs complexity, but it's too much for such a small feature.

I'll be honest, I searched through dozens of github repos that use webtorrent and none of them use this feature. I also asked around on reddit - no answer so far. If it was up to me, I'd remove it in a second. We're changing the select/deselect behaviour anyway, why not get rid of this unused feature in exchange for sensible selec/deselect? But I understand that we should avoid breaking changes, even for such small things.

I'm willing to properly test the current implementation, which should be as close to the original implementation as we can get, though it is complex.

---

<details>
<summary>
The other path we can consider (besides removing and the current implementation) is that we could change the behaviour slightly (dropdown)
</summary>

Before the current solution we simply stored 1 notification callback per selection and just called that. That was simple and understandable, but I think the issue would be the following:

```js
torrent.select(0, 20, 0, () => console.log('downloading [0 - 20]'))
torrent.select(10, 40, 0, () => console.log('downloading [10 - 40]'))
```
The second selection truncates the first one, so it's kinda false that it's "downloading [0 - 20]", since it's actually "downloading [0 - 9]".

Logging is pretty much the only legit use case for the notify function, so at least that should be accurate.

So what if we use this simple solution, but we pass the `from` and `to` values to `notify`?

```js
torrent.select(0, 20, 0, (from, to) => console.log(`downloading [${from} - ${to}]`))
```

This would always report it accurately and it would keep the implementation "stupid simple". I think it's worth considering.

---

My issue with merging selections is that it would be even less accurate than not doing anything. If we would merge notifications, it would call the `[0-20]` notification even when the downloaded pieces are actually between 21-40. I think in that case the "no-change" solution might be more correct, since it would never "lie", it would just not call between 10-20.

</details>

ThaUnknown commented:
I agree that I'd also like to drop this feature, it was so unusable that I created a `verified` even because it's simply easier to work with, and a lot faster, but unfortunately we can't do that, because I was told that breaking functionality should be a major release, and a major release needs to be agreed upon by everyone, this doesn't warrant that.

That said we can simply neuter it, as I mentioned merging callbacks, instead of adding this `from` `to` stuff would be a lot easier, be a lot safer, and sure in a few cases it would be called when not necessary, but it's still called rarely enough where that doesn't matter, and since the callback doesn't give any data, the developer needs to check the bitfield if the piece he's looking for is verified, so the developer's end functionality wouldn't change

detarkende committed:
fix: merge notification functions and simplify selection code

detarkende commented:
@ThaUnknown I took your advice and went with the merging solution. I am however thinking about splitting the inserted selection at the intersections and only merging the notify function of the section that's actually overlapping. It's just a thought for now, but that would mean that the notify function merging would be much more precise, though at the expense of creating more objects in the `_items` array. 

Let me know what you think. I can provide some sketches if that helps.

ThaUnknown commented on lib/selections.js:
```diff
@@ -53,28 +65,26 @@ export class Selections {
-            isCoveringExisting(range, notification)
-      ) {
-        notification.notify()
+    newItem.notify = newItem.notify ?? noop
```
don't noop the notifiy, if it doesn't exist, skip it, less function calls

detarkende replied:
Got it. removed the noop, instead the notify functions are called like `s.notify?.()`.

ThaUnknown commented on lib/selections.js:
```diff
@@ -53,28 +65,26 @@ export class Selections {
+    if (!newItem.isStreamSelection) {
+      const { notify: oldNotify } = newItem
+      // Merge notifications of intersecting items into the new item's notify function
+      const intersectingNotifyFunctions = new Set(
```
is there any point in de-duplicating them? it's unlikely it will ever happen and it forces you to iterate over the items 3 times, instead of once

detarkende replied:
It was only there to merge all the noops, since they would all point to the same reference. But since noops are removed, I replaced this with a single `for..of` loop.

detarkende committed:
fix: remove noop from selections notify

ThaUnknown commented on lib/selections.js:
```diff
@@ -0,0 +1,185 @@
+    }
+    if (!newItem.isStreamSelection) {
+      const { notify: oldNotify } = newItem
+      // Merge notifications of intersecting items into the new item's notify function
```
shouldn't this entire thing be skipped if notify doesn't exist on the item or am I missunderstanding?

detarkende replied:
i don't think so. even though the inserted item doesn't have a notify function, if the merged items do, then we should merge those into the inserted one.

ThaUnknown commented on lib/selections.js:
```diff
@@ -0,0 +1,185 @@
+      if (intersectingNotifyFunctions.length > 0) {
+        newItem.notify = () => {
+          intersectingNotifyFunctions.forEach(fn => fn())
+          oldNotify()
```
this will call undefined

detarkende committed:
fix: optional chaining for `oldNotify` call

SilentBot1 commented:
Hey, just after some further testing from my side:
```javascript
const { default: WebTorrent } = await import("./index.js") 
const client = new WebTorrent()
client.add("magnet:?xt=urn:btih:08ada5a7a6183aae1e09d831df6748d566095a10&dn=Sintel&tr=udp%3A%2F%2Fexplodie.org%3A6969&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969&tr=udp%3A%2F%2Ftracker.empire-js.us%3A1337&tr=udp%3A%2F%2Ftracker.leechers-paradise.org%3A6969&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337&tr=wss%3A%2F%2Ftracker.btorrent.xyz&tr=wss%3A%2F%2Ftracker.fastcast.nz&tr=wss%3A%2F%2Ftracker.openwebtorrent.com&ws=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2F&xs=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel.torrent", {deselect: true})
let server = client.createServer()
server.listen(8443)
```
And running the command:
```bash
curl http://localhost:8443/webtorrent/08ada5a7a6183aae1e09d831df6748d566095a10/Sintel/Sintel.mp4 -O
```
Appears to create two stream selections:
![image](https://github.com/webtorrent/webtorrent/assets/15065825/678a4d76-950f-429a-886c-3d68363f4a5b)

Both selections are updated / destroyed as the stream progresses/closes:
![image](https://github.com/webtorrent/webtorrent/assets/15065825/ad99e7dd-a0ee-4add-8479-01ddd7c7c2e7)

This was the only thing I've found, everything else looks good and appears to work as expected / designed and I'll be happy to approve once the above is resolved.

detarkende commented:
@SilentBot1 - Thanks for the review. I tried to reproduce your situation, but I have to say, I don't know what I'm missing, but it only inserts 1 selection for me.

This is my script:
```js
import WebTorrent from './index.js'
import MemoryChunkStore from 'memory-chunk-store'

const client = new WebTorrent()
client.add('magnet:?xt=urn:btih:08ada5a7a6183aae1e09d831df6748d566095a10&dn=Sintel&tr=udp%3A%2F%2Fexplodie.org%3A6969&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969&tr=udp%3A%2F%2Ftracker.empire-js.us%3A1337&tr=udp%3A%2F%2Ftracker.leechers-paradise.org%3A6969&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337&tr=wss%3A%2F%2Ftracker.btorrent.xyz&tr=wss%3A%2F%2Ftracker.fastcast.nz&tr=wss%3A%2F%2Ftracker.openwebtorrent.com&ws=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2F&xs=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel.torrent', { deselect: true, store: MemoryChunkStore })
const server = client.createServer()
server.listen(8443)

/** @type {import('./lib/torrent.js').default} */
const torrent = client.torrents[0]

setInterval(() => {
  console.log(torrent._selections)
}, 1000)
```

And this is the output for me, when I run your `curl` command in a separate terminal instance:

<img width="392" alt="image" src="https://github.com/webtorrent/webtorrent/assets/64217783/f29e7a60-c726-49fc-99ea-b87b40af2a8b">

Can you please try it again? I have rebased the repo now, but that shouldn't have an effect on this (theoretically).

I tested this with node 16, 18 and 20. Which version are you using? If you have the time, we could check in a video call or something, but I'm just out of ideas on how to reproduce this 😄

SilentBot1 commented:
Hey @detarkende, I'm now also unable to replicate this (on Node v18.16.0) after the rebase and a `git reset --hard detarkende/2755-non-overlapping-selections`, so I'm assuming this was a skill issue on my half, but this now looks all good to go.

detarkende commented:
Thank you @ThaUnknown and @SilentBot1! Let's ship this thing 🚀 

@ThaUnknown, to me it shows that you have 2 unresolved threads, but I cannot find them. Could you please check and maybe reapprove?

detarkende commented:
Thanks everyone! It says that only those with write access to webtorrent can merge this PR, so from my end if either of you have this permission, we can merge this 😄 🚀

ThaUnknown commented:
it will be merged when I have the time.

ThaUnknown commented:
yeah add that 1 test, and I'll merge this, this is defo an A tier pull request

detarkende committed:
feat: add test for complex select-deselect behaviour to check for leaks

detarkende commented on test/node/client-deselect.js:
```diff
@@ -0,0 +1,109 @@
+      t.equal(torrent.pieces.filter((a) => a === null).length, (torrent.pieces.length - 1 - lastPieceIndex))
+    }
+  })
+})
```
@ThaUnknown I just pushed the test you asked for, but please WAIT. I might have found an issue.

detarkende replied:
@ThaUnknown I believe I have fixed the issue that I found this morning. pushed the code just now. All tests passed in my local plus I added a few new ones, so I think we're pretty much done.

detarkende committed:
fix: insideExisting handling for edge cases and add tests

ThaUnknown commented:
good PR, well documented, clean simple code, includes typings for static analysis

ThaUnknown commented:
@detarkende if you're interested in working on other things, in order of priority:
- #2465
- #2594 had a pull request, but it was 50% finished, and abandoned
- #2685
- #2082
- #2467
- #2581

I wouldn't work on types, webtorrent needs to be re-written to use async/await instead of callbacks [this has been on the roadman for years], which will break all the types anyways

ThaUnknown commented:
you can fix that, by re-writing this to use byte ranges internally, not pieces :)

detarkende commented:
@ThaUnknown yep, I'll take a look at it this week. I might have to rewrite these tests altogether because they are kind of weird anyway.
